ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java

```diff
@@ -102,29 +105,57 @@
   private final KsqlEngineMetrics engineMetrics;
   private final ScheduledExecutorService aggregateMetricsCollector;
   private final FunctionRegistry functionRegistry;
-  private SchemaRegistryClient schemaRegistryClient;
+  private final SchemaRegistryClient schemaRegistryClient;
 
-  public KsqlEngine(final KsqlConfig ksqlConfig, final KafkaTopicClient topicClient) {
+  public KsqlEngine(
+      final KsqlConfig ksqlConfig
+  ) {
     this(
         ksqlConfig,
-        topicClient,
-        new KsqlSchemaRegistryClientFactory(ksqlConfig).create(),
+        new KsqlSchemaRegistryClientFactory(ksqlConfig).create()
+    );
+  }
+
+  public KsqlEngine(
+      final KsqlConfig ksqlConfig,
+      final SchemaRegistryClient schemaRegistryClient
+  ) {
+    this(
+        ksqlConfig,
+        new KafkaTopicClientImpl(ksqlConfig.getKsqlAdminClientConfigProps()),
+        true,
+        schemaRegistryClient,
         new MetaStoreImpl()
     );
   }
 
   public KsqlEngine(
       final KsqlConfig ksqlConfig,
       final KafkaTopicClient topicClient,
+      final SchemaRegistryClient schemaRegistryClient
+  ) {
+    this(
+        ksqlConfig,
+        topicClient,
+        false,
+        schemaRegistryClient,
+        new MetaStoreImpl()
+    );
+  }
+
+  KsqlEngine(
+      final KsqlConfig ksqlConfig,
+      final KafkaTopicClient topicClient,
+      final boolean topicClientOwned,
```

Comment:
This is just so we know if to close the topicClient? Seems off to me.

Reply:
I've refactored to remove the need for it. (Though it still requires a slightly hacky test-only derived class gain access to the protected constructor. This is mainly as I don't want to expose public constructors that take in a topicClient, as doing so then means the engine might not _own_ the topic client again).